<?xml version="1.0"?>

<launch>

    <arg name="model" default="husky"/>
    <arg name="init_pose" default="-x 0.00 -y 0.00 -Y 0.0"/>
    <arg name="laser_enabled" default="$(optenv HUSKY_LMS1XX_ENABLED true)"/>
    <arg name="realsense_enabled" default="$(optenv HUSKY_REALSENSE_ENABLED true)"/>
    <arg name="urdf_extras" default="$(optenv HUSKY_URDF_EXTRAS)"/>
    <arg name="ipx" default="0.00"/>
    <arg name="ipy" default="0.00"/>
    <arg name="navigating" default="true"/>

    <param name="robot_description" command="$(find xacro)/xacro $(find dissertation_simulation)/xacro/husky.urdf.xacro robot_namespace:=$(arg model) tf_prefix:=$(arg model)_tf laser_enabled:=$(arg laser_enabled) realsense_enabled:=$(arg realsense_enabled) urdf_extras:=$(arg urdf_extras)"/>
    <node pkg="gazebo_ros" type="spawn_model" name="spawn_model" args="-urdf -param /$(arg model)/robot_description -model $(arg model) $(arg init_pose)" output="screen"/>


    <rosparam file="$(find dissertation_simulation)/config/control.yaml" command="load"/>
    <!-- <rosparam param="husky_velocity_controller/base_frame_id" subst_value="True">$(arg model)_tf/base_link</rosparam> -->
    <rosparam param="husky_velocity_controller/base_frame_id" subst_value="True">$(arg model)_tf/base_footprint</rosparam>
    <!-- <param name="/$(arg model)/controller_spawner/husky_velocity_controller/base_frame_id" value="$(arg model)_tf/base_link"/> -->
    <!-- <param name="husky_velocity_controller/base_frame_id" value="$(arg model)_tf/base_link"/> -->

    <node pkg="controller_manager" type="spawner" name="controller_spawner" respawn="true" output="screen" args="husky_joint_publisher husky_velocity_controller">
        <!-- <param name="tf_prefix" value="$(arg model)_tf"/> -->
        <!-- <param name="base_frame_id" value="$(arg model)_tf/base_link"/> -->
        <param name="base_frame_id" value="base_footprint"/>
    </node>

    <rosparam param="wheel_radius_multiplier" ns="husky_velocity_controller" subst_value="True">
        $(optenv HUSKY_WHEEL_MULTIPLIER 1.0)
    </rosparam>

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" output="screen">
        <rosparam command="load" file="$(find dissertation_simulation)/config/localization.yaml"/>
        <!-- <param name="map_frame" value="/$(arg model)/slam_map"/> -->
        <param name="base_link_frame" value="base_footprint"/>
        <!-- <rosparam param="base_link_frame" subst_value="True">$(arg model)_tf/base_link</rosparam> -->
        <!-- <rosparam param="odom_frame" subst_value="True">$(arg model)_tf/odom</rosparam> -->
    </node>
    <!-- <param name="$(arg model)/ekf_localization/base_link_frame" value="$(arg model)_tf/base_link"/> -->
    <!-- <param name="$(arg model)/ekf_localization/odom_frame" value="$(arg model)_tf/odom"/> -->
    
    
      <node pkg="interactive_marker_twist_server" type="marker_server" name="twist_marker_server" output="screen">
        <!-- <param name="link_name" value="$(arg model)_tf/base_link" /> -->
        <param name="link_name" value="base_footprint" />
      </node>
    
      <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen">
        <param name="tf_prefix" value="$(arg model)_tf"/>
      </node>
    
      <node pkg="twist_mux" type="twist_mux" name="twist_mux" output="screen">
        <rosparam command="load" file="$(find husky_control)/config/twist_mux.yaml" />
        <remap from="cmd_vel_out" to="husky_velocity_controller/cmd_vel"/>
      </node>

      <!-- <node pkg="fake_localization" type="fake_localization" name="fake_localization">
        <param name="odom_frame_id" value="$(arg model)_tf/odom" />
        <param name="delta_x" value="0" />
        <param name="delta_y" value="0" />
        <param name="delta_yaw" value="0" />
        <param name="base_frame_id" value="$(arg model)_tf/base_link" />
      </node> -->
    
    <include file="$(find dissertation_simulation)/launch/gmapping.launch" unless="$(arg navigating)">
      <arg name="model" value="$(arg model)"/>
    </include>

    <!-- <include file="$(find husky_gazebo)/launch/realsense.launch"/>
    <param name="target_frame" value="$(arg model)_tf/base_link"/> -->

    <!-- <remap from="/scan" to="/model/scan"/> -->

    <!-- <node pkg="tf" type="static_transform_publisher" name="map_odom_broadcaster" args="0 0 0 0 0 0 /map /$(arg model)_tf/odom 100"/> -->
    <node pkg="tf" type="static_transform_publisher" name="map_model_broadcaster" args="0 0 0 0 0 0 /map /$(arg model)_tf/map 100"/>
    <!-- <node pkg="tf" type="static_transform_publisher" name="base_link_broadcaster" args="0 0 0 0 0 0 /$(arg model)_tf/base_link /$(arg model)_tf/base_footprint 100"/> -->

    <!-- <remap from="base_link" to="$(arg model)_tf/base_link"/> -->
    <!-- <remap from="odom" to="$(arg model)_tf/odom"/> -->

    <include file="$(find dissertation_simulation)/launch/move_base.launch" >
        <arg name="model" value="$(arg model)"/>
        <arg name="navigating" value="$(arg navigating)"/>
        <arg name="no_static_map" value="true" unless="$(arg navigating)"/>
    </include>

    <include file="$(find dissertation_simulation)/launch/amcl.launch" if="$(arg navigating)">
        <arg name="model" value="$(arg model)"/>
        <arg name="ipx" value="$(arg ipx)"/>
        <arg name="ipy" value="$(arg ipy)"/>
    </include>

    <include file="$(find dissertation_simulation)/launch/explore.launch" unless="$(arg navigating)">
      <arg name="model" value="$(arg model)"/>
    </include>

    <!-- <arg name="joy_dev" default="$(optenv HUSKY_JOY_DEVICE /dev/input/js0)" />
  
    <rosparam command="load" file="$(find dissertation_simulation)/config/teleop_ps4.yaml" />
  
    <node pkg="joy" type="joy_node" name="joy_node" />
  
    <node pkg="teleop_twist_joy" type="teleop_node" name="teleop_twist_joy"/> -->

    

    <include file="$(find dissertation_simulation)/launch/continuous_detection.launch">
      <arg name="model" value="$(arg model)"/>
    </include>

</launch>